### PR TITLE
Add data, ui lattices and ui primitives for the discuss context

### PR DESCRIPTION
**Pull request by rtshkmr**
## Motivation
This PR defines some functions that are used when defining the two lattices that shall be kept within the socket -- data lattice and ui_lattice. 

Additionally, I've also added some bottom up ui primitives that we will be building on top of for the next bit when displaying discussions.

Note the following:
1. I'm currently looking only at the :index action for discussions
2. there's no pagination for now, pagination shall be left for later -- until the various moving parts get merged together

## Screenshots 
image

image

### CODE REVIEW

File: lib/vyasa/sangh/sheaf_lattice.ex [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    |> Enum.flat_map(fn level ->
+      root_sheafs
+      |> Enum.map(fn sheaf -> to_string(sheaf.path) end)
+      |> Enum.flat_map(fn sheaf_id ->
+        Sangh.get_child_sheafs_by_session(sangh_id, sheaf_id, level)
+      end)
+      |> Enum.map(fn s ->
+        {s.path.labels,
+         s
+         |> Repo.preload(:marks)}
```
Comment (rtshkmr) [inline lib/vyasa/sangh/sheaf_lattice.ex:33]: QQ: @ks0m1c is it okay to preload marks when creating the sheaf?

for now, i'm creating a complete sheaf lattice, while ignoring the pagination piece. Will ignore pagination until things are more ready.

File: docs/crudrice.livemd [hunk 2]
```diff
@@ -216,6 +219,174 @@ root_sheafs
 
 ```
 
+## Sheaf Lattice Definition
```
Comment (rtshkmr) [inline docs/crudrice.livemd:222]: To get the meat of this PR, referring to the livebook from this section onwards should be sufficient.

File: lib/vyasa/sangh/sheaf_lattice.ex [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+  # Fetch immediate children based on particular parent
+  # Fetch immediate children of a specific level 0 node:
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 1, ["cf27deab", nil])
+
+  # Fetch immediate children of a specific level 1 node:
+      iex> SheafLattice.read_sheaf_lattice(sheaf_lattice, 2, ["c9cbcb0c", "65c1ac0c", nil])
+  """
+
+  def read_sheaf_lattice(%{} = sheaf_lattice, level \\ 0, match \\ nil) do
```
Comment (rtshkmr) [inline lib/vyasa/sangh/sheaf_lattice.ex:73]: Here's the main accessor helper function to read from the lattice. My next PR shall be relying heavily on this. 

The examples above show the various cases I expect this read function to be used for.

File: lib/vyasa_web/components/contexts/ui_state.ex [hunk 4]
```diff
@@ -4,6 +4,48 @@
 # VyasaWeb.Context.Components module.                                           #
 #################################################################################
 
+defmodule VyasaWeb.Context.Components.UiState.Sheaf do
```
Comment (rtshkmr) [inline lib/vyasa_web/components/contexts/ui_state.ex:7]: the ui state for sheaf is a container that encapsulates the ui state of a collection of marks

**Review by rtshkmr — COMMENTED** [review]
This PR defines some functions that are used when defining the two lattices that shall be kept within the socket -- data lattice and ui_lattice

Note the following:
1. I'm currently looking only at the :index action for discussions
2. there's no pagination for now, pagination shall be left for later -- until the various moving parts get merged together

File: lib/vyasa/sangh/sheaf_lattice.ex (hunk 1 above)
Comment (ks0m1c) [reply lib/vyasa/sangh/sheaf_lattice.ex:33]: should preload at get_root_sheafs_by_session/1

File: lib/vyasa/sangh/sheaf_lattice.ex (hunk 3 above)
Comment (ks0m1c) [reply lib/vyasa/sangh/sheaf_lattice.ex:73]: HAHAHA this filter pattern is quite cool create_sheaf_lattice_filter could easily use 
for bot get_in and get_and_update_in behav

[Access.filter(&(create_sheaf_lattice_filter(&1)))]

File: lib/vyasa_web/components/contexts/components.ex [hunk 5]
```diff
@@ -390,28 +390,39 @@ defmodule VyasaWeb.Context.Components do
   end
 
   @doc """
-  TODO: implement a reddit-top-comment-like UI for this
   A brief view of a sheaf, showing the contextually relevant information about it.
+  Shows a barebones, brief summary of a sheaf, which gives just enough context for someone to
```
Comment (ks0m1c) [inline lib/vyasa_web/components/contexts/components.ex:394]: probably should be sheaf body as title and first mark

File: lib/vyasa/sangh/sheaf_lattice.ex (hunk 3 above)
Comment (rtshkmr) [reply lib/vyasa/sangh/sheaf_lattice.ex:71]: I'll carry on first, will leave it to you to make this change yeah

**Comment by rtshkmr** [comment]
@ks0m1c give an approval if this is ready to merge, i want to merge the ready PRs in 

I don't intent to commit anything new to this

File: lib/vyasa/sangh/sheaf_lattice.ex (hunk 3 above)
Comment (ks0m1c) [reply lib/vyasa/sangh/sheaf_lattice.ex:71]: you want me to commit this in(?)

File: lib/vyasa/sangh/sheaf_lattice.ex (hunk 1 above)
Comment (rtshkmr) [reply lib/vyasa/sangh/sheaf_lattice.ex]: done this, marking this as resolved

File: lib/vyasa_web/components/contexts/components.ex (hunk 5 above)
Comment (rtshkmr) [reply lib/vyasa_web/components/contexts/components.ex:387]: you're right, will get back to this when i'm doing ui iteration again after the current immediate next set of tasks

**Review by ks0m1c — APPROVED** [review]

**Pull request merged**